src/backend/renderer/element/memory.rs

```diff
@@ -0,0 +1,278 @@
+
+        match self.textures.entry(texture_id) {
+            Entry::Occupied(entry) => {
+                if !buffer_damage.is_empty() {
+                    trace!(log, "updating memory with damage {:#?}", &buffer_damage);
+                    renderer.update_memory(entry.get().downcast_ref().unwrap(), &self.mem, buffer_damage)?
+                }
+            }
+            Entry::Vacant(entry) => {
+                trace!(log, "importing memory");
+                let tex = renderer.import_memory(&self.mem, buffer_size, false)?;
+                entry.insert(Box::new(tex));
+            }
+        };
+
+        self.renderer_seen.insert(texture_id, current_commit);
+
+        Ok(self
+            .textures
+            .get(&texture_id)
+            .unwrap()
+            .downcast_ref::<<R as Renderer>::TextureId>()
+            .unwrap())
+    }
+}
+
+/// A memory backed render buffer
+#[derive(Debug, Clone)]
+pub struct MemoryRenderBuffer {
+    id: Id,
+    inner: Arc<Mutex<MemoryRenderBufferInner>>,
+}
+
+impl Default for MemoryRenderBuffer {
+    fn default() -> Self {
+        Self {
+            id: Id::new(),
+            inner: Default::default(),
+        }
+    }
+}
+
+impl MemoryRenderBuffer {
+    /// Initialize a [`MemoryRenderBuffer`] from existing memory
+    pub fn from_memory(mem: &[u8]) -> Self {
+        let inner = MemoryRenderBufferInner::from_memory(mem);
+        MemoryRenderBuffer {
+            id: Id::new(),
+            inner: Arc::new(Mutex::new(inner)),
+        }
+    }
+
+    /// Render to the memory buffer
+    pub fn render(&self) -> RenderContext<'_> {
+        let guard = self.inner.lock().unwrap();
+        RenderContext {
+            buffer: guard,
+            damage: Vec::new(),
+        }
+    }
+
+    fn current_commit(&self) -> CommitCounter {
+        self.inner.lock().unwrap().damage_tracker.current_commit()
+    }
+
+    fn damage_since(&self, commit: Option<CommitCounter>) -> Option<Vec<Rectangle<i32, Buffer>>> {
+        self.inner.lock().unwrap().damage_tracker.damage_since(commit)
+    }
+}
+
+/// A render context for [`MemoryRenderBuffer`]
+#[derive(Debug)]
+pub struct RenderContext<'a> {
+    buffer: MutexGuard<'a, MemoryRenderBufferInner>,
+    damage: Vec<Rectangle<i32, Buffer>>,
+}
+
+impl<'a> RenderContext<'a> {
+    /// Resize the buffer
+    pub fn resize(&mut self, size: usize) {
+        self.buffer.resize(size)
+    }
+
+    /// Draw to the buffer
+    pub fn draw<F>(&mut self, f: F)
+    where
+        F: FnOnce(&mut [u8]) -> Vec<Rectangle<i32, Buffer>>,
+    {
+        let draw_damage = f(&mut self.buffer.mem);
+        self.damage.extend(draw_damage);
+    }
+}
+
+impl<'a> Drop for RenderContext<'a> {
+    fn drop(&mut self) {
+        self.buffer.damage_tracker.add(&self.damage);
+    }
+}
+
+/// A render element for [`MemoryRenderBuffer`]
+#[derive(Debug)]
+pub struct MemoryRenderBufferRenderElement {
```

Comment:
Getting some serious Java-vibes from these names :joy: 
(But I don't necessarily have a better idea and it is descriptive at least.)